src/data/W/endofunctor.lean

```diff
@@ -0,0 +1,172 @@
+/-
+Copyright (c) 2022 Joseph Hua. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joseph Hua
+-/
+import data.W.basic
+import category_theory.endofunctor.algebra
+import category_theory.equivalence
+import category_theory.functor.category
+
+/-!
+# Endofunctors associated to `W_type`s
+
+Any `W_type` can be seen as initial algebras of their associated polynomial endofunctors.
+
+
+## Main results
+* `W_type.endofunctor`: makes the endofunctor associated to the `W_type`
+* `W_type.data`: combines the data of `α : Type _` and `β : α → Type _` into a structure.
+  This is given an instance of a category, which can be seen as a subcategory of `Type _ ⥤ Type _`
+* `W_type.data.endofunctor`: packages `W_type.endofunctor` into a (fully faithful) functor from
+  `W_type.data` to the endofunctor category `Type _ ⥤ Type _`
+* `W_type.algebra` makes a `W_type` an algbera over its associated endofunctor
+* `W_type.is_initial` shows that `W_type.algebra` is an initial algebra
+-/
+
+universes u₀ u₁
+
+namespace W_type
+
+open category_theory
+open category_theory.endofunctor
+
+/-- The polynomial endofunctor associated to a `W_type` -/
+@[simps] def endofunctor {α : Type u₀} (β : α → Type u₁) : Type (max u₀ u₁) ⥤ Type (max u₀ u₁) :=
+{ obj := λ X, Σ (a : α), β a → X,
+  map := λ X Y f p, ⟨ p.1 , f ∘ p.2 ⟩ }
+
+/--
+The data of a `W_type` consists of an index `α` of constructors and
+their dependently indexed arities `β`.
+This will form a category.
+-/
+@[nolint check_univs]
+structure data :=
```

Comment:
I think data is not a good name. Something mentioning that it's a polynomial functor would be clearer. You could also use the same definition as in https://github.com/leanprover-community/mathlib/blob/master/src/data/pfunctor/univariate/basic.lean

Reply:
Should I just be importing this file? That seems a bit incorrect though, the definition `pfunctor` in that file should be somewhere in `data.W.basic` and then both my file and this `pfunctor` file should import this definition

Reply:
`pfunctor` is useful in more contexts than just W types. You can see in `pfunctor`, there is the construction for M and W types and the generalization of M and W types to multivariate polynomial functors. You can also look in https://github.com/leanprover-community/mathlib/tree/master/src/data/qpf/multivariate/constructions for more constructions when we combine polynomial functors with quotient types.